File: apina-core/src/test/java/fi/evident/apina/output/ts/CodeWriterTest.java
package fi.evident.apina.output.ts;

import org.junit.Test;

import java.util.LinkedHashMap;
import java.util.Map;

import static java.util.Arrays.asList;
import static java.util.Collections.emptyMap;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

public class CodeWriterTest {

    private final CodeWriter writer = new CodeWriter();

    @Test
    public void indents() {
        writer.writeLine("foo ");
        writer.writeBlock(() -> {
            writer.writeLine("bar");
            writer.writeLine("baz");
        });
        writer.writeLine();

        assertThat(writer.getOutput(), is("foo \n{\n    bar\n    baz\n}\n"));
    }

    @Test
    public void stringValues() {
        writer.writeValue("foo 'bar' baz\nquux");

        assertThat(writer.getOutput(), is("'foo \\'bar\\' baz\\nquux'"));
    }

    @Test
    public void numberValues() {
        writer.writeValue(42);

        assertThat(writer.getOutput(), is("42"));
    }

    @Test
    public void booleanValues() {
        writer.writeValue(false);

        assertThat(writer.getOutput(), is("false"));
    }

    @Test
    public void mapValues() {
        Map<String,Integer> map = new LinkedHashMap<>();
        map.put("foo", 1);
        map.put("bar", 2);
        map.put("baz", 3);
        writer.writeValue(map);

        assertThat(writer.getOutput(), is("{\n    'foo': 1,\n    'bar': 2,\n    'baz': 3\n}"));
    }

    @Test
    public void emptyMapValues() {
        writer.writeValue(emptyMap());

        assertThat(writer.getOutput(), is("{}"));
    }

    @Test
    public void collectionValues() {
        writer.writeValue(asList(1, 2, 3));

        assertThat(writer.getOutput(), is("[1, 2, 3]"));
    }
}

File: apina-core/src/test/kotlin/fi/evident/apina/output/ts/CodeWriterTest.kt
package fi.evident.apina.output.ts

import org.junit.jupiter.api.Test
import java.util.Collections.emptyMap
import kotlin.test.assertEquals

class CodeWriterTest {

    private val writer = CodeWriter()

    @Test
    fun indents() {
        writer.writeLine("foo")
        writer.writeBlock {
            writer.writeLine("bar")
            writer.writeLine("baz")
        }

        assertEquals("""
            foo
            {
                bar
                baz
            }
        """.trimIndent(), writer.output)
    }

    @Test
    fun `indents are properly handled when text has line breaks`() {
        writer.writeLine("foo")
        writer.writeBlock {
            writer.writeLine("""
                bar
                baz
            """.trimIndent())
        }

        assertEquals("""
            foo
            {
                bar
                baz
            }
        """.trimIndent(), writer.output)
    }

    @Test
    fun stringValues() {
        writer.writeValue("foo 'bar' baz\nquux")

        assertEquals("'foo \\'bar\\' baz\\nquux'", writer.output)
    }

    @Test
    fun numberValues() {
        writer.writeValue(42)

        assertEquals("42", writer.output)
    }

    @Test
    fun booleanValues() {
        writer.writeValue(false)

        assertEquals("false", writer.output)
    }

    @Test
    fun mapValues() {
        writer.writeValue(mapOf(
                "foo" to 1,
                "bar" to 2,
                "baz" to 3))

        assertEquals("""
            {
                'foo': 1,
                'bar': 2,
                'baz': 3
            }
        """.trimIndent(), writer.output)
    }

    @Test
    fun emptyMapValues() {
        writer.writeValue(emptyMap<Any, Any>())

        assertEquals("{}", writer.output)
    }

    @Test
    fun collectionValues() {
        writer.writeValue(listOf(1, 2, 3))

        assertEquals("[1, 2, 3]", writer.output)
    }
}
